apps/examples/reboot_reason_test/reboot_reason_main.c

```diff
@@ -0,0 +1,234 @@
+#include <sys/prctl.h>
+#include <tinyara/reboot_reason.h>
+
+#define REBOOT_REASON_TEST_VAL 30
+
+static sem_t wait_sem;
+
+static void display_reboot_reason_option(void)
+{
+	printf("\nSelect Reboot Reason Test Option.\n");
+	printf("\t-Press H or h : HW Reset Test\n");
+#ifdef CONFIG_MM_ASSERT_ON_FAIL
+	printf("\t-Press M or m : Memory Allocation Fail Test\n");
+#endif
+#ifdef CONFIG_WATCHDOG
+	printf("\t-Press W or w : Watchdog Reset Test\n");
+#endif
+	printf("\t-Press V or v : Random Value Write-Read Test\n");
+	printf("\t-Press C or c : Clear reboot reason\n");
+	printf("\t-Press R or r : Check previous reboot reason\n");
+	printf("\t-Press X or x : Terminate Tests.\n");
+}
+#ifdef CONFIG_WATCHDOG
+static void watchdog_test(void)
+{
+	int fd;
+	int ret;
+	/* Test for Watchdog Reset */
+	printf("After watchdog reset, Expected Reboot reason is %d\n", REBOOT_SYSTEM_WATCHDOG);
+
+	fd = open(CONFIG_WATCHDOG_DEVPATH, O_RDWR);
+	if (fd < 0) {
+		printf("ERROR: Fail to open %s\n", CONFIG_WATCHDOG_DEVPATH);
+		return;
+	}
+
+	ret = ioctl(fd, WDIOC_SETTIMEOUT, 1000);
+	if (ret != OK) {
+		printf("ERROR: Fail to set watchdog timeout.\n");
+		close(fd);
+		return;
+	}
+
+	ret = ioctl(fd, WDIOC_START, 0);
+	if (ret != OK) {
+		printf("ERROR: Fail to start watchdog.\n");
+		close(fd);
+		return;
+	}
+
+	/* Wait until watchdog timeout */
+	sem_wait(&wait_sem);
+}
+#endif
+
+#ifdef CONFIG_MM_ASSERT_ON_FAIL
+static void memory_alloc_fail_test(void)
+{
+	struct mallinfo data;
+	int largest_size;
+	int *alloc_ptr;
+	/* Test for Memory Allocation Fail */
+	printf("After Memory Allocation Fail, Expected Reboot reason is %d\n", REBOOT_SYSTEM_MEMORYALLOCFAIL);
+
+#ifdef CONFIG_CAN_PASS_STRUCTS
+	data = mallinfo();
+#else
+	mallinfo(&data);
+#endif
+	largest_size = data.mxordblk;
+	/* Try to allocate not available size. */
+	alloc_ptr = (int *)malloc(largest_size + 1);
+
+	/* This line cannot be reached,
+	 * because there will be assert on memory allocation failure from malloc.
+	 */
+	free(alloc_ptr);
+}
+#endif
+
+static void reboot_board(void)
+{
+#ifdef CONFIG_LIB_BOARDCTL
+	printf("Board will be reset automatically.\n");
+	boardctl(BOARDIOC_RESET, 0);
+#else
+	printf("Please reboot the board manually.\n");
+#endif
+}
+
+/****************************************************************************
+ * reboot_reason_main
+ ****************************************************************************/
+
+#ifdef CONFIG_BUILD_KERNEL
+int main(int argc, FAR char *argv[])
+#else
+int reboot_reason_main(int argc, char *argv[])
+#endif
+{
+	int ch;
+	bool is_testing = true;
+
+	printf("Reboot Reason Test Start!\n");
+
+	sem_init(&wait_sem, 0, 0);
+	while (is_testing) {
+		display_reboot_reason_option();
+		ch = getchar();
+		switch (ch) {
+		case 'H':
+		case 'h':
+			printf("Expected reboot reason after hw reset is : %d\n", REBOOT_SYSTEM_HW_RESET);
+			printf("Press HW Reset Button.\n");
+			/* Wait for pressing the hw reset button. */
+			sem_wait(&wait_sem);
```

Comment:
Do we need to use additional sem to wait? We can use just while loop, then we can save memory even though it's enough small.